cinder/scheduler/filters/shard_filter.py

```diff
@@ -186,42 +180,28 @@ def _filter_by_k8s_cluster(self, backends, filter_properties):
         if availability_zone:
             query_filters = {'availability_zone': availability_zone}
 
-        query_metadata = {CSI_CLUSTER_METADATA_KEY: cluster_name}
-        k8s_volumes = VolumeList.get_all_by_metadata(
-            cinder_context.get_admin_context(),
-            project_id, query_metadata,
+        k8s_hosts_vols = db.get_hosts_by_volume_metadata(
+            key=CSI_CLUSTER_METADATA_KEY,
+            value=cluster_name,
             filters=query_filters)
 
-        if not k8s_volumes:
-            return backends
-
-        k8s_hosts = set(volume_utils.extract_host(v.host, 'host')
-                        for v in k8s_volumes
-                        if v.id != spec.get('volume_id') and v.host)
-        if not k8s_hosts:
+        if not k8s_hosts_vols:
             return backends
 
-        def _backend_shards(backend_state):
-            cap = backend_state.capabilities.get(self._CAPABILITY_NAME)
-            return cap.split(',') if cap else []
-
-        hosts_shards_map = {
-            volume_utils.extract_host(bs.host, 'host'):
-                _backend_shards(bs)
-            for bs in self.all_backend_states}
-
-        k8s_shards = set()
-        for host in k8s_hosts:
-            shards = hosts_shards_map[host]
-            k8s_shards.update(shards)
-
         return [
             b for b in backends if
             (not self._is_vmware(b) or
-             set(_backend_shards(b)) & k8s_shards)
+             volume_utils.extract_host(b.host, 'host') in k8s_hosts_vols)
```

Comment:
Here I am not sure, I think the idea was only to return the host with the most volumes.
Possibly, that also could be done directly in the query as well.

Reply:
I am myself not quite clear on what is better, the query already should return only one host in most cases, and in rare cases maybe a handful. So the advantage of doing it in SQL over python is rather small.
Having it done in the query limits the use-cases for exactly this one.

After having slept over it, I would say, just return here the most dominant one only.
But then the question is, how do we do migrations of a k8s cluster out of a vcenter?